### ISSUE TITLE
Allowing `typing.NoReturn` as a return hint for `__init__` methods

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#10342]
**Feature**

Allow [`typing.NoReturn`](https://docs.python.org/3/library/typing.html#typing.NoReturn) as a return hint for `__init__` methods, instead of constraining it to just `None`.

**Pitch**

The following is valid Python code:

```python
class A:
    def __init__(self):
        raise RuntimeError
```
```python
>>> A()
Traceback (most recent call last):
    ...
RuntimeError
```

Many projects using [`Boost.Python`](https://github.com/boostorg/python) will employ its option to define classes exported to Python which *aren't meant to be instantiated directly in Python*, but rather via helper functions/methods, and attempts to instantiate these classes directly in Python will look like the above. These are created by defining a [`boost::python::class`](https://github.com/boostorg/python/blob/f5d14ef15e98838b3cf18692d46c481287ed50d0/include/boost/python/class.hpp#L587-L599) with [`boost::python::no_init`](https://github.com/boostorg/python/blob/f5d14ef15e98838b3cf18692d46c481287ed50d0/include/boost/python/class.hpp#L59), and interfacing with the Python API typically looks like the following:

```python
>>> import MyBoostPythonModule  # A binary library
>>>
>>> # The final ancestor is `builtins.object`, just like standard Python classes.
>>> # Verify that this is `True`
>>> MyBoostPythonModule.MyNoInitClass.__mro__
(MyBoostPythonModule.MyNoInitClass, Boost.Python.instance, object)
>>> MyBoostPythonModule.MyNoInitClass.__mro__[-1] is object
True
>>>
>>> # Bad attempt at instantiation
>>> # See https://github.com/boostorg/python/blob/f5d14ef15e98838b3cf18692d46c481287ed50d0/src/object/class.cpp#L656-L674
>>> MyBoostPythonModule.MyNoInitClass()
Traceback (most recent call last):
    ...
RuntimeError: This class cannot be instantiated from Python
>>>
>>> # Example instance-getting function
>>> type(MyBoostPythonModule.GetMyNoInitClass).__mro__
(Boost.Python.function, object)
>>> MyBoostPythonModule.GetMyNoInitClass()  # OK
<MyBoostPythonModule.MyNoInitClass at 0x7fc56b2f25d0>
```

Having the ability to define stub classes which look like

```python
from typing import NoReturn

class A:
    def __init__(self) -> NoReturn: ...
```

and to be able to use `mypy` to detect bad attempts at creating objects would greatly aid bug detection.

**Comment by JelleZijlstra** [comment]
This also recently came up in https://github.com/python/typeshed/pull/5137#discussion_r609791521. Annotating `__init__` as `NoReturn` seems like a reasonable way to indicate that a class cannot be instantiated.

**Comment by AlexWaygood** [comment]
Fixed in #13480 🎉
